src/components/views/rooms/RoomList.js

```diff
@@ -344,7 +339,26 @@ module.exports = React.createClass({
         lists["im.vector.fake.archived"] = [];
 
         const dmRoomMap = DMRoomMap.shared();
-        MatrixClientPeg.get().getRooms().forEach((room) => {
+
+        // If there are any tags selected, constrain the rooms listed to the
+        // visible rooms as determined by this._visibleRooms. Here, we
+        // de-duplicate and filter out rooms that the client doesn't know
+        // about (hence the Set and the null-guard on `room`).
+        let rooms = [];
+        if (this.state.selectedTags.length > 0) {
+            const roomSet = new Set();
+            this._visibleRooms.forEach((roomId) => {
+                const room = MatrixClientPeg.get().getRoom(roomId);
+                if (room) {
+                    roomSet.add(room);
+                }
+            });
+            rooms = Array.from(roomSet);
```

Comment:
Would it be more sensible to do this de-duping & validating when we make `_visibleRooms` so we only have to do it once? In fact, would it make sense to just use `_visibleRooms` as the authoritative list of rooms that are visible whether or not filters are in place? (presumably moving it to state if so)

Reply:
Yep, that would make sense. Not sure why we'd move it to state though, given that we're using `lists` as the state that gets rendered (and we'd deriving state from state).